example/apps/test_security/tests/celery_log.py

```diff
@@ -0,0 +1,723 @@
+
+from security.backends.signals import (
+    celery_task_invocation_started, celery_task_invocation_expired, celery_task_invocation_triggered,
+    celery_task_invocation_ignored, celery_task_invocation_timeout, celery_task_run_started, celery_task_run_failed,
+    celery_task_run_retried, celery_task_run_succeeded, celery_task_run_output_updated
+)
+
+from apps.test_security.tasks import error_task, retry_task, sum_task, unique_task, ignored_after_success_task
+
+from .base import (
+    BaseTestCaseMixin, _all_, TRUNCATION_CHAR, assert_equal_dict_data, set_signal_receiver, test_call_command,
+    _not_none_
+)
+
+
+@override_settings(SECURITY_BACKENDS={})
+class CeleryLogTestCase(BaseTestCaseMixin, ClientTestCase):
+
+    @data_consumer('create_user')
+    def test_sum_celery_task_should_be_logged(self, user):
+        expected_invocation_started_data = {
+            'name': 'sum_task',
+            'queue_name': 'default',
+            'input': '5, 8',
+            'task_args': [5, 8],
+            'task_kwargs': {},
+            'applied_at': _not_none_,
+            'is_async': False,
+            'is_unique': False,
+            'is_on_commit': False,
+            'start': _not_none_,
+        }
+        expected_invocation_triggered_data = {
+            **expected_invocation_started_data,
+            'triggered_at': _not_none_,
+            'stale_at': None,
+            'estimated_time_of_first_arrival': _not_none_,
+            'expires_at': None,
+            'celery_task_id': _not_none_,
+            'is_duplicate': False,
+        }
+
+        expected_run_started_data = {
+            'name': 'sum_task',
+            'queue_name': None,
+            'input': '5, 8',
+            'task_args': [5, 8],
+            'task_kwargs': {},
+            'start': _not_none_,
+            'retries': 0,
+            'celery_task_id': _not_none_,
+        }
+        expected_run_succeeded_data = {
+            **expected_run_started_data,
+            'stop': _not_none_,
+            'result': 13
+        }
+
+        with set_signal_receiver(celery_task_invocation_started,
+                                 expected_invocation_started_data) as invocation_started_receiver, \
+                set_signal_receiver(celery_task_invocation_triggered,
+                                    expected_invocation_triggered_data) as invocation_triggered_receiver, \
+                set_signal_receiver(celery_task_run_started,
+                                    expected_run_started_data) as run_started_receiver, \
+                set_signal_receiver(celery_task_run_succeeded,
+                                    expected_run_succeeded_data) as run_succeeded_receiver, \
+                set_signal_receiver(celery_task_run_output_updated) as run_output_updated_receiver, \
+                set_signal_receiver(celery_task_run_failed) as run_failed_receiver, \
+                set_signal_receiver(celery_task_run_retried) as run_retried_receiver:
```

Comment:
Huh, this is sort of clumsy. Wouldn't be better to just create some helper class, that would automatically register to all these signals, and have internal counters that would increment when signal called. Or something light that. I believe this repeated initialization (signal registering) is not necessary. Something like
```
with TestSignalReceiver() as receiver:
    ... # do your stuff
    assert_equal(receiver.calls['celery_task_run_succeeded'], 1)
```
or
```
assert_equal(receiver.calls, {
    'invocation_started_receiver' : 1,
    'run_output_updated_receiver' : 6,
    ...
})
```
The second way of asserting is even better, because it will fail if any unexpected signal is fired, so you don't have to explicitly assert signals that are not supposed to be sent. (the dict will only contain signals that were fired at least once)

Reply:
yes you are right. I will try to use the decorator which I wrote for project log testing and yes I should rewrite the set_signal_receiver to this decorator. Thanks